castai/resource_node_template.go

```diff
@@ -221,8 +223,15 @@ func resourceNodeTemplate() *schema.Resource {
 						FieldNodeTemplateStorageOptimized: {
 							Type:        schema.TypeBool,
 							Optional:    true,
-							Default:     false,
-							Description: "Storage optimized instance constraint - will only pick storage optimized nodes if true",
+							Description: "Storage optimized instance constraint - will only pick storage optimized nodes if true (deprecated).",
+							Deprecated:  "`storage_optimized` attribute will be removed in the next major release. Use `storage_optimized_status` instead.",
+						},
+						FieldNodeTemplateStorageOptimizedStatus: {
+							Type:             schema.TypeString,
+							Optional:         true,
+							Default:          "",
+							Description:      "Storage optimized instance constraint - will only pick storage optimized nodes if enabled and won't pick if disabled. Empty value will have no effect. Supported values: `enabled`, `disabled` or empty string.",
+							ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice([]string{"", Enabled, Disabled}, false)),
```

Comment:
Did you have a chance to look at `ExactlyOnOf`? This would improve UX by displaying clear error if someone tries to configure both old and new fields at the same time.
```suggestion
							ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice([]string{"", Enabled, Disabled}, false)),
							ExactlyOneOf: []string{fmt.Sprint(FieldNodeTemplateConstraints, ".0.", FieldNodeTemplateStorageOptimized), fmt.Print(FieldNodeTemplateConstraints, ".0.", FieldNodeTemplateStorageOptimizedStatus)},
```

Similar thing could be done for the other pair of fields.

Reply:
as we are defaulting to empty string, we will always get value from `FieldNodeTemplateStorageOptimizedStatus`, therefore this constraint will always fire if `FieldNodeTemplateStorageOptimized` is set